File: src/photo_workflow/sharpness.py
"""FR-1.4: Focus scoring via Tenengrad + SML with subject-aware regions."""

from __future__ import annotations

import logging
from pathlib import Path

import click
import cv2
import numpy as np

from .scoring_types import SharpnessScores, SubjectContext

logger = logging.getLogger(__name__)

_SUPPORTED_EXTS = {
    ".jpg", ".jpeg", ".png", ".tiff", ".tif", ".raw",
    ".cr2", ".cr3", ".nef", ".arw", ".dng",
}

# Thresholds for blur classification
_SHARP_THRESHOLD = 0.25  # Normalized score above this = "sharp"
_RATIO_BOKEH_THRESHOLD = 3.0  # subject/background ratio above this = bokeh
_MOTION_ANISOTROPY_THRESHOLD = 2.5  # gradient-direction imbalance => motion, not defocus


def _tenengrad(gray: np.ndarray) -> float:
    """Compute Tenengrad focus measure (Sobel gradient magnitude squared).

    Returns raw unnormalized score (higher = sharper).
    """
    if gray.size == 0:
        return 0.0
    gx = cv2.Sobel(gray, cv2.CV_64F, 1, 0, ksize=3)
    gy = cv2.Sobel(gray, cv2.CV_64F, 0, 1, ksize=3)
    tenengrad = float(np.mean(gx**2 + gy**2))
    return tenengrad


def _sml(gray: np.ndarray) -> float:
    """Compute Sum of Modified Laplacian focus measure.

    Returns raw unnormalized score (higher = sharper).
    """
    if gray.size == 0:
        return 0.0
    img = gray.astype(np.float64)
    # Modified Laplacian: |d2f/dx2| + |d2f/dy2| (avoids sign cancellation)
    lxx = cv2.Sobel(img, cv2.CV_64F, 2, 0, ksize=3)
    lyy = cv2.Sobel(img, cv2.CV_64F, 0, 2, ksize=3)
    ml = np.abs(lxx) + np.abs(lyy)
    return float(np.mean(ml))


def _laplacian_variance(gray: np.ndarray) -> float:
    """Compute Laplacian variance (traditional focus measure).

    Returns raw unnormalized score (higher = sharper).
    Useful for pure checkerboard patterns where Sobel derivatives are zero.
    """
    if gray.size == 0:
        return 0.0
    lap = cv2.Laplacian(gray, cv2.CV_64F)
    return float(lap.var())


def _focus_maps(gray: np.ndarray) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """Per-pixel Tenengrad and modified-Laplacian maps (+ raw gradients)."""
    gx = cv2.Sobel(gray, cv2.CV_64F, 1, 0, ksize=3)
    gy = cv2.Sobel(gray, cv2.CV_64F, 0, 1, ksize=3)
    ten_map = gx * gx + gy * gy
    img = gray.astype(np.float64)
    lxx = cv2.Sobel(img, cv2.CV_64F, 2, 0, ksize=3)
    lyy = cv2.Sobel(img, cv2.CV_64F, 0, 2, ksize=3)
    ml_map = np.abs(lxx) + np.abs(lyy)
    return ten_map, ml_map, gx, gy


def _gradient_anisotropy(gx: np.ndarray, gy: np.ndarray, region: np.ndarray) -> float:
    """Ratio of the stronger to the weaker gradient direction in a region.

    Motion blur suppresses gradients along the motion direction, so a
    directionally smeared region scores high; defocus is isotropic (~1).
    """
    if not region.any():
        return 1.0
    ex = float((gx[region] ** 2).mean())
    ey = float((gy[region] ** 2).mean())
    lo = min(ex, ey)
    if lo <= 0.0:
        return 1.0
    return max(ex, ey) / lo


def _region_score_from_maps(
    ten_map: np.ndarray, ml_map: np.ndarray, gray: np.ndarray, mask: np.ndarray
) -> float:
    """Geometric mean of Tenengrad and SML averaged over the masked pixels.

    Averaging over the region's own pixels (not the whole frame) keeps the
    score independent of the region's area — the old whole-frame mean over a
    zeroed copy made a tack-sharp subject at 5% of frame score ~5% of its
    true value and get misclassified as misfocused/motion blur. The mask is
    eroded by the Sobel kernel radius so genuine subject/background
    transition edges don't inflate the region's score.
    """
    if mask.sum() == 0:
        return 0.0

    eroded = cv2.erode(mask.astype(np.uint8), np.ones((3, 3), np.uint8))
    region = eroded > 0
    if not region.any():
        region = mask > 0

    ten = float(ten_map[region].mean())
    sml_val = float(ml_map[region].mean())
    if ten <= 0.0 and sml_val <= 0.0:
        # 3x3 Sobel's transverse smoothing cancels period-2 checkerboards
        # entirely; the raw Laplacian (no smoothing) still sees them.
        lap = np.abs(cv2.Laplacian(gray, cv2.CV_64F))
        raw = float(lap[region].mean())
        if raw <= 0.0:
            return 0.0
    elif ten <= 0.0 or sml_val <= 0.0:
        raw = max(ten, sml_val)
    else:
        raw = float(np.sqrt(ten * sml_val))

    # Normalize by mean luminance of the region to handle exposure differences
    mean_luma = float(gray[mask > 0].mean())
    luma_factor = max(mean_luma, 1.0) / 128.0

    # Empirical normalization constant (tuned for typical photos)
    normalized = raw / (500.0 * luma_factor)
    return min(normalized, 1.0)


def _compute_region_sharpness(gray: np.ndarray, mask: np.ndarray) -> float:
    """Region sharpness in ~[0, 1]; see _region_score_from_maps."""
    ten_map, ml_map, _, _ = _focus_maps(gray)
    return _region_score_from_maps(ten_map, ml_map, gray, mask)


def _classify_blur(
    subject_sharp: bool,
    bg_sharp: bool,
    ratio: float,
    subject_anisotropy: float = 1.0,
) -> str:
    """Classify blur type based on region sharpness analysis.

    Args:
        subject_sharp: Whether subject region exceeds sharpness threshold.
        bg_sharp: Whether background region exceeds sharpness threshold.
        ratio: subject_sharpness / background_sharpness.
        subject_anisotropy: Directional gradient imbalance of the subject
            region; high values mean directionally smeared (moved) rather
            than uniformly defocused.

    Returns:
        One of: "sharp", "bokeh", "motion_subject", "motion_global", "misfocused"
    """
    if subject_sharp and bg_sharp:
        return "sharp"
    if subject_sharp and not bg_sharp:
        return "bokeh"  # Sharp subject over soft background = intentional DOF
    if not subject_sharp and bg_sharp:
        # A directional smear means the subject moved during exposure; an
        # isotropic softness means focus missed. (This branch previously
        # always returned "misfocused", leaving motion_subject unreachable.)
        if subject_anisotropy >= _MOTION_ANISOTROPY_THRESHOLD:
            return "motion_subject"
        return "misfocused"
    # Neither sharp
    return "motion_global"


def score_sharpness_detailed(ctx: SubjectContext) -> SharpnessScores:
    """Compute subject-aware sharpness scores using Tenengrad + SML.

    Args:
        ctx: SubjectContext containing image_gray and subject_mask.

    Returns:
        SharpnessScores with per-region scores and blur classification.
    """
    gray = ctx.image_gray
    mask = ctx.subject_mask
    inv_mask = 1 - mask

    # Compute the gradient maps once; all three regions read from them.
    ten_map, ml_map, gx, gy = _focus_maps(gray)
    subject_score = _region_score_from_maps(ten_map, ml_map, gray, mask)
    background_score = _region_score_from_maps(ten_map, ml_map, gray, inv_mask)

    # Eye region sharpness
    eye_score = subject_score  # Default: same as subject
    if ctx.faces:
        # Use first face's eye landmarks
        face = ctx.faces[0]
        if "left_eye" in face.landmarks and "right_eye" in face.landmarks:
            le = face.landmarks["left_eye"]
            re = face.landmarks["right_eye"]
            # Create eye region mask (rectangle around both eyes)
            eye_mask = np.zeros_like(gray, dtype=np.uint8)
            eye_w = abs(re[0] - le[0])
            eye_h = max(eye_w // 3, 10)
            y_center = (le[1] + re[1]) // 2
            x_min = min(le[0], re[0]) - eye_w // 4
            x_max = max(le[0], re[0]) + eye_w // 4
            y_min = y_center - eye_h
            y_max = y_center + eye_h
            # Clamp to image bounds
            y_min = max(0, y_min)
            y_max = min(gray.shape[0], y_max)
            x_min = max(0, x_min)
            x_max = min(gray.shape[1], x_max)
            if y_max > y_min and x_max > x_min:
                eye_mask[y_min:y_max, x_min:x_max] = 1
                eye_score = _region_score_from_maps(ten_map, ml_map, gray, eye_mask)
    elif ctx.detections:
        # For animals: use upper third of primary detection bbox as eye region
        for det in ctx.detections:
            if det.class_name in ("cat", "dog", "bird", "horse", "bear"):
                x, y, w, h = det.bbox
                head_h = h // 3
                eye_mask = np.zeros_like(gray, dtype=np.uint8)
                y_min = max(0, y)
                y_max = min(gray.shape[0], y + head_h)
                x_min = max(0, x)
                x_max = min(gray.shape[1], x + w)
                if y_max > y_min and x_max > x_min:
                    eye_mask[y_min:y_max, x_min:x_max] = 1
                    eye_score = _region_score_from_maps(ten_map, ml_map, gray, eye_mask)
                break

    # Sharpness contrast ratio
    sharpness_contrast = subject_score / max(background_score, 0.01)

    # Blur classification
    subject_sharp = subject_score >= _SHARP_THRESHOLD
    bg_sharp = background_score >= _SHARP_THRESHOLD
    anisotropy = _gradient_anisotropy(gx, gy, mask > 0)
    blur_type = _classify_blur(subject_sharp, bg_sharp, sharpness_contrast, anisotropy)

    # Overall score (weighted combination favoring subject and eye)
    overall = 0.5 * subject_score + 0.35 * eye_score + 0.15 * min(sharpness_contrast / 5.0, 1.0)
    overall = min(overall, 1.0)

    return SharpnessScores(
        subject=round(subject_score, 4),
        eye_region=round(eye_score, 4),
        background=round(background_score, 4),
        sharpness_contrast=round(sharpness_contrast, 4),
        blur_type=blur_type,
        overall=round(overall, 4),
    )


# --- Backward compatibility ---

# Legacy threshold for the old API
BLUR_THRESHOLD = 100.0


def score_sharpness(path: Path) -> float:
    """Compute sharpness score for a single image file.

    Backward-compatible wrapper that returns a single float in [0.0, 1.0].
    Uses Tenengrad + SML geometric mean, with Laplacian variance as a fallback.
    """
    from .raw_loader import load_gray

    try:
        image_gray = load_gray(path)
    except Exception:
        logger.warning("Could not load image for sharpness: %s", path)
        return 0.0

    # Compute focus measures on full image (no masking to avoid edge artifacts)
    ten = _tenengrad(image_gray)
    sml_val = _sml(image_gray)

    if ten <= 0 and sml_val <= 0:
        # Fallback to Laplacian variance (e.g., for perfect checkerboards)
        lap_var = _laplacian_variance(image_gray)
        normalized = min(lap_var / (BLUR_THRESHOLD * 10.0), 1.0)
        return round(normalized, 4)

    if ten > 0 and sml_val > 0:
        # Geometric mean of both measures
        raw = float(np.sqrt(ten * sml_val))
    else:
        # Use whichever is non-zero
        raw = max(ten, sml_val)

    # Normalize by mean luminance
    mean_luma = float(image_gray.mean())
    luma_factor = max(mean_luma, 1.0) / 128.0

    # Normalize to [0, 1] range
    normalized = raw / (500.0 * luma_factor)
    normalized = min(normalized, 1.0)

    return round(normalized, 4)


@click.command("sharpness")
@click.argument("path", type=click.Path(exists=True, path_type=Path))
@click.option(
    "--threshold",
    default=BLUR_THRESHOLD,
    show_default=True,
    help="Blur threshold for normalization (variance units).",
)
def main(path: Path, threshold: float) -> None:
    """Score sharpness for PATH (file or directory of images)."""
    targets = sorted(path.iterdir()) if path.is_dir() else [path]
    for p in targets:
        if not (p.is_file() and p.suffix.lower() in _SUPPORTED_EXTS):
            continue
        score = score_sharpness(p)
        click.echo(f"{p}: {score:.4f}")

File: android/core/src/main/kotlin/com/photonforge/core/cv/Sharpness.kt
package com.photonforge.core.cv

import com.photonforge.core.FaceDetection
import com.photonforge.core.GrayImage
import com.photonforge.core.Mask
import com.photonforge.core.ObjectDetection
import com.photonforge.core.SharpnessScores
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/** Port of photo_workflow.sharpness (FR-1.4): Tenengrad + SML, subject-aware regions. */
object Sharpness {

    const val SHARP_THRESHOLD = 0.25
    const val MOTION_ANISOTROPY_THRESHOLD = 2.5

    private val ANIMAL_EYE_CLASSES = setOf("cat", "dog", "bird", "horse", "bear")

    class FocusMaps(val ten: FloatArray, val ml: FloatArray, val gx: FloatArray, val gy: FloatArray)

    fun focusMaps(gray: GrayImage): FocusMaps {
        val gx = ImageOps.sobelX(gray)
        val gy = ImageOps.sobelY(gray)
        val ten = FloatArray(gx.size)
        for (i in gx.indices) ten[i] = gx[i] * gx[i] + gy[i] * gy[i]
        val lxx = ImageOps.sobelXX(gray)
        val lyy = ImageOps.sobelYY(gray)
        val ml = FloatArray(lxx.size)
        for (i in lxx.indices) ml[i] = abs(lxx[i]) + abs(lyy[i])
        return FocusMaps(ten, ml, gx, gy)
    }

    private fun maskedMean(map: FloatArray, mask: Mask): Double {
        var s = 0.0
        var n = 0L
        for (i in map.indices) {
            if (mask.data[i].toInt() != 0) { s += map[i]; n++ }
        }
        return if (n > 0) s / n else 0.0
    }

    fun regionScoreFromMaps(maps: FocusMaps, gray: GrayImage, mask: Mask): Double {
        if (mask.sum() == 0L) return 0.0

        var region = ImageOps.erode3(mask)
        if (region.sum() == 0L) region = mask

        val ten = maskedMean(maps.ten, region)
        val smlVal = maskedMean(maps.ml, region)
        val raw: Double
        if (ten <= 0.0 && smlVal <= 0.0) {
            val lap = ImageOps.laplacian(gray)
            var s = 0.0
            var n = 0L
            for (i in lap.indices) if (region.data[i].toInt() != 0) { s += abs(lap[i]); n++ }
            val rawLap = if (n > 0) s / n else 0.0
            if (rawLap <= 0.0) return 0.0
            raw = rawLap
        } else if (ten <= 0.0 || smlVal <= 0.0) {
            raw = max(ten, smlVal)
        } else {
            raw = sqrt(ten * smlVal)
        }

        var lumaSum = 0.0
        var lumaN = 0L
        for (i in gray.pixels.indices) {
            if (mask.data[i].toInt() != 0) { lumaSum += gray.pixels[i]; lumaN++ }
        }
        val meanLuma = if (lumaN > 0) lumaSum / lumaN else 0.0
        val lumaFactor = max(meanLuma, 1.0) / 128.0
        return min(raw / (500.0 * lumaFactor), 1.0)
    }

    fun gradientAnisotropy(gx: FloatArray, gy: FloatArray, mask: Mask): Double {
        var ex = 0.0
        var ey = 0.0
        var n = 0L
        for (i in gx.indices) {
            if (mask.data[i].toInt() != 0) {
                ex += gx[i].toDouble() * gx[i]
                ey += gy[i].toDouble() * gy[i]
                n++
            }
        }
        if (n == 0L) return 1.0
        ex /= n
        ey /= n
        val lo = min(ex, ey)
        if (lo <= 0.0) return 1.0
        return max(ex, ey) / lo
    }

    fun classifyBlur(subjectSharp: Boolean, bgSharp: Boolean, anisotropy: Double): String = when {
        subjectSharp && bgSharp -> "sharp"
        subjectSharp && !bgSharp -> "bokeh"
        !subjectSharp && bgSharp ->
            if (anisotropy >= MOTION_ANISOTROPY_THRESHOLD) "motion_subject" else "misfocused"
        else -> "motion_global"
    }

    fun score(
        gray: GrayImage,
        mask: Mask,
        faces: List<FaceDetection> = emptyList(),
        detections: List<ObjectDetection> = emptyList(),
    ): SharpnessScores {
        val maps = focusMaps(gray)
        val invMask = mask.inverted()
        val subjectScore = regionScoreFromMaps(maps, gray, mask)
        val backgroundScore = regionScoreFromMaps(maps, gray, invMask)

        var eyeScore = subjectScore
        val face = faces.firstOrNull()
        if (face != null && face.landmarks.containsKey("left_eye") && face.landmarks.containsKey("right_eye")) {
            val le = face.landmarks.getValue("left_eye")
            val re = face.landmarks.getValue("right_eye")
            val eyeW = abs(re[0] - le[0])
            val eyeH = max(eyeW / 3, 10)
            val yCenter = (le[1] + re[1]) / 2
            var xMin = min(le[0], re[0]) - eyeW / 4
            var xMax = max(le[0], re[0]) + eyeW / 4
            var yMin = yCenter - eyeH
            var yMax = yCenter + eyeH
            yMin = max(0, yMin); yMax = min(gray.height, yMax)
            xMin = max(0, xMin); xMax = min(gray.width, xMax)
            if (yMax > yMin && xMax > xMin) {
                eyeScore = regionScoreFromMaps(maps, gray, rectMask(gray, xMin, yMin, xMax, yMax))
            }
        } else if (faces.isEmpty() && detections.isNotEmpty()) {
            for (det in detections) {
                if (det.className in ANIMAL_EYE_CLASSES) {
                    val (x, y, w, h) = listOf(det.bbox[0], det.bbox[1], det.bbox[2], det.bbox[3])
                    val headH = h / 3
                    val yMin = max(0, y)
                    val yMax = min(gray.height, y + headH)
                    val xMin = max(0, x)
                    val xMax = min(gray.width, x + w)
                    if (yMax > yMin && xMax > xMin) {
                        eyeScore = regionScoreFromMaps(maps, gray, rectMask(gray, xMin, yMin, xMax, yMax))
                    }
                    break
                }
            }
        }

        val sharpnessContrast = subjectScore / max(backgroundScore, 0.01)
        val anisotropy = gradientAnisotropy(maps.gx, maps.gy, mask)
        val blurType = classifyBlur(
            subjectScore >= SHARP_THRESHOLD,
            backgroundScore >= SHARP_THRESHOLD,
            anisotropy,
        )
        var overall = 0.5 * subjectScore + 0.35 * eyeScore + 0.15 * min(sharpnessContrast / 5.0, 1.0)
        overall = min(overall, 1.0)

        return SharpnessScores(
            subject = round4(subjectScore),
            eyeRegion = round4(eyeScore),
            background = round4(backgroundScore),
            sharpnessContrast = round4(sharpnessContrast),
            blurType = blurType,
            overall = round4(overall),
        )
    }

    private fun rectMask(gray: GrayImage, xMin: Int, yMin: Int, xMax: Int, yMax: Int): Mask {
        val data = ByteArray(gray.width * gray.height)
        for (y in yMin until yMax) {
            for (x in xMin until xMax) data[y * gray.width + x] = 1
        }
        return Mask(gray.width, gray.height, data)
    }
}

internal fun round4(v: Double): Double = Math.round(v * 10000.0) / 10000.0
